### PR TITLE
Add SegNetBasic model

### PR DESCRIPTION
**Pull request by mitmul**
Merge after #132 

- SegNetBasic model is from this Caffe impl.: https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Models/segnet_basic_train.prototxt

### Training

```
cd examples/segnet
CHAINER_TYPE_CHECK=0 python -OO -W ignore train.py --gpu 0
```
`eval_semantic_segmentation` makes a lot of warnings, so I suppressed them by giving `-W ignore` option.

### Memo

- According to the original implementation, the authors seemed to perform LR flipping to the input images for data augmentation: https://github.com/alexgkendall/caffe-segnet/blob/segnet-cleaned/src/caffe/layers/dense_image_data_layer.cpp#L168-L175
- Chainer's LRN layer is different from Caffe's one in terms of the meaning of "alpha" argment: https://github.com/alexgkendall/caffe-segnet/blob/segnet-cleaned/src/caffe/layers/lrn_layer.cpp#L121

## Results

The evaluation result after 16000 iterations:
```
                    Sky : 0.8675
               Building : 0.6487
                   Pole : 0.1795
           Road_marking : 0.8493
                   Road : 0.6230
               Pavement : 0.6051
                   Tree : 0.1813
             SignSymbol : 0.1901
              Fence,Car : 0.5924
             Pedestrian : 0.2402
              Bicyclist : 0.2111
==================================
               mean IoU : 0.4716
 Class average accuracy : 0.6705
Global average accuracy : 0.8266
```

```
Global accuracy : 82.8%
Class accuracy : 62.3%
Mean IoU : 46.3%
```

The trained weights to replicate the same results as below is here: [model_iteration-16000](https://www.dropbox.com/s/exas66necaqbxyw/model_iteration-16000?dl=1)

### CODE REVIEW

**Comment by rezoo** [comment]
It seems that this PR contains ``.sync-config.cson`` that contains internal information. Please remove it and rebase the branch.

**Comment by mitmul** [comment]
@rezoo Oops, sorry. I removed it and rebased this branch. Please check it again.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import glob
+import os.path as osp
```
Comment (rezoo) [inline chainercv/datasets/camvid/camvid_dataset.py]: `import numpy as np` is allowed, but `import os.path as osp` is not allowed. Please use `os.path` directly.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    def __init__(self, data_dir='auto', mode='train'):
+        if mode not in ['train', 'val', 'test']:
+            raise ValueError(
+                'Please pick mode from \'train\', \'val\', \'test\'')
+
+        if data_dir == 'auto':
+            data_dir = get_camvid()
+
+        img_list_fn = osp.join(data_dir, '{}.txt'.format(mode))
```
Comment (rezoo) [inline chainercv/datasets/camvid/camvid_dataset.py]: How about using `filename` and `filenames` instead of `fn` and `fns`?

File: chainercv/links/model/segnet/segnet_basic.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import chainer
+import chainer.functions as F
+import chainer.links as L
+
+
+class SegNetBasic(chainer.Chain):
+
```
Comment (rezoo) [inline chainercv/links/model/segnet/segnet_basic.py]: Please add the description, and add it to `.rst` files.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    .. _`u`: https://github.com/alexgkendall/SegNet-Tutorial/tree/master/CamVid
+
+    Args:
+        data_dir (string): Path to the root of the training data. If this is
+            :obj:`auto`, this class will automatically download data for you
+            under :obj:`$CHAINER_DATASET_ROOT/pfnet/chainercv/camvid`.
+        mode ({'train', 'val', 'test'}): Select from dataset splits used
+            in VOC.
+
+    """
```
Comment (rezoo) [inline chainercv/datasets/camvid/camvid_dataset.py]: Please add this class to the `.rst` file.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            tuple of color image and label whose shapes are (3, H, W) and
+            (1, H, W) respectively. H and W are height and width of the
+            images. The dtype of the color image is :obj:`numpy.float32` and
+            the dtype of the label image is :obj:`numpy.int32`.
+
+        """
+        if i >= len(self):
+            raise IndexError('index is too large')
+        image_fn, label_fn = self.fns[i]
+        image = read_image(image_fn, color=True)
```
Comment (Hakuyume) [inline chainercv/datasets/camvid/camvid_dataset.py]: `image` -> `img` (please refer to #159)

File: chainercv/datasets/camvid/camvid_dataset.py (hunk 1 above)
Comment (yuyu2172) [inline chainercv/datasets/camvid/camvid_dataset.py]: Right. I fixed other datasets #174.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    """Dataset class for a semantic segmantion task on CamVid `u`_.
+
+    .. _`u`: https://github.com/alexgkendall/SegNet-Tutorial/tree/master/CamVid
+
+    Args:
+        data_dir (string): Path to the root of the training data. If this is
+            :obj:`auto`, this class will automatically download data for you
+            under :obj:`$CHAINER_DATASET_ROOT/pfnet/chainercv/camvid`.
+        mode ({'train', 'val', 'test'}): Select from dataset splits used
```
Comment (yuyu2172) [inline chainercv/datasets/camvid/camvid_dataset.py]: Can you use `split` instead of `mode`?
This kwarg has recently been changed in other datasets. #165.

File: chainercv/datasets/camvid/camvid_dataset.py (hunk 2 above)
Comment (yuyu2172) [inline chainercv/datasets/camvid/camvid_dataset.py]: Actually, fn is used in other part of ChainerCV (e.g. `datasets/cub/cub_utils`).

File: chainercv/links/model/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from chainercv.links.model.segnet.segnet_basic import SegNetBasic  # NOQA
```
Comment (yuyu2172) [inline chainercv/links/model/__init__.py]: Can you move imports to `chainercv.links.__init__`?
Also, losses should be moved there as well.

File: examples/segnet/evaluate.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+serializers.load_npz(args.snapshot, model)
+if args.gpu >= 0:
+    model.to_gpu(args.gpu)
+
+test = CamVidDataset(mode='test')
+
+n_positive = [0 for _ in range(args.n_class)]
+n_true = [0 for _ in range(args.n_class)]
+n_true_positive = [0 for _ in range(args.n_class)]
+for i in tqdm(range(0, len(test), args.batchsize)):
```
Comment (yuyu2172) [inline examples/segnet/evaluate.py]: please do not use tqdm

File: chainercv/datasets/camvid/camvid_dataset.py (hunk 2 above)
Comment (Hakuyume) [inline chainercv/datasets/camvid/camvid_dataset.py]: I prefer `filename` to `fn`. How about modifying existing datasets?

Comment (yuyu2172) [inline chainercv/datasets/camvid/camvid_dataset.py]: OK. I am fine with `filename`.

File: examples/segnet/train.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+# Dataset
+train = CamVidDataset(mode='train')
+
+
+def transform(in_data):
+    x, t = in_data
+    if np.random.rand() > 0.5:
+        x = x[:, :, ::-1]
+        t = t[:, ::-1]
+    return x, t
```
Comment (yuyu2172) [inline examples/segnet/train.py]: `x, t` --> `img, label`.
This is the naming convention used in ChainerCV.

File: chainercv/links/loss/semantic_segmentation_loss.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    y = F.argmax(F.softmax(y), axis=1)
+    if y.ndim == 3:
+        y = y[:, None, :, :]
+    if t.ndim == 3:
+        t = t[:, None, :, :]
+    return [np.mean(ret)
+            for ret in eval_semantic_segmentation(y.data, t.data, n_class)]
+
+
+class PixelwiseSigmoidLoss(chainer.Chain):
```
Comment (yuyu2172) [inline chainercv/links/loss/semantic_segmentation_loss.py]: Can you add docs and tests for these classes?

File: chainercv/links/loss/semantic_segmentation_loss.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def segmentation_accuracies(y, t, n_class):
+    y = F.argmax(F.softmax(y), axis=1)
+    if y.ndim == 3:
+        y = y[:, None, :, :]
+    if t.ndim == 3:
+        t = t[:, None, :, :]
+    return [np.mean(ret)
+            for ret in eval_semantic_segmentation(y.data, t.data, n_class)]
```
Comment (yuyu2172) [inline chainercv/links/loss/semantic_segmentation_loss.py]: I am suspecting that this fails when cupy array is fed.

File: chainercv/links/loss/semantic_segmentation_loss.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    def __init__(self, model, n_class, calc_accuracy=True):
+        super(PixelwiseSigmoidLoss, self).__init__(predictor=model)
+        self.n_class = n_class
+        self.calc_accuracy = calc_accuracy
+
+    def __call__(self, x, t):
+        self.y = self.predictor(x)
+        self.loss = F.sigmoid_cross_entropy(self.y, t)
+        reporter.report({'loss': self.loss}, self)
+        if self.calc_accuracy:
```
Comment (yuyu2172) [inline chainercv/links/loss/semantic_segmentation_loss.py]: How about `compute_accuracy` instead of `calc_accuracy`.
This is the name used in Chainer's links/model/classifier.py

File: chainercv/links/loss/semantic_segmentation_loss.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+import chainer.functions as F
+from chainer import reporter
+
+import numpy as np
+
+from chainercv.evaluations import eval_semantic_segmentation
+
+
+def segmentation_accuracies(y, t, n_class):
+    y = F.argmax(F.softmax(y), axis=1)
```
Comment (yuyu2172) [inline chainercv/links/loss/semantic_segmentation_loss.py]: This function can be private because it will only be used inside this file.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+            (1, H, W) respectively. H and W are height and width of the
+            images. The dtype of the color image is :obj:`numpy.float32` and
+            the dtype of the label image is :obj:`numpy.int32`.
+
+        """
+        if i >= len(self):
+            raise IndexError('index is too large')
+        image_fn, label_fn = self.filenames[i]
+        img = read_image(image_fn, color=True)
+        label_map = read_image(label_fn, dtype=np.int32, color=False)[0]
```
Comment (yuyu2172) [inline chainercv/datasets/camvid/camvid_dataset.py]: Please return labels by (1, H, W).
This is the shape used in VOCSemanticSegmentationDataset.

This shape is selected because all the image related transforms under `chainercv/transforms/` assumes that images are 3 three dimensional.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+                'Please pick mode from \'train\', \'val\', \'test\'')
+
+        if data_dir == 'auto':
+            data_dir = get_camvid()
+
+        img_list_filename = os.path.join(data_dir, '{}.txt'.format(mode))
+        self.filenames = [
+            [os.path.join(data_dir, fn.replace('/SegNet/CamVid/', ''))
+             for fn in line.split()] for line in open(img_list_fn)]
+
```
Comment (yuyu2172) [inline chainercv/datasets/camvid/camvid_dataset.py]: img_list_fn does not exist

File: examples/segnet/evaluate.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    img, lbl = concat_examples(test[i:i + args.batchsize], args.gpu)
```
Comment (yuyu2172) [inline examples/segnet/evaluate.py]: `img, lbl` --> `img, label`.

File: examples/segnet/train.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+from chainer import iterators
+from chainer import optimizers
+from chainer import training
+from chainer.training import extensions
+from chainercv.datasets import CamVidDataset
+from chainercv.datasets import TransformDataset
+from chainercv.links.loss import PixelwiseSoftmaxLossWithWeight
+from chainercv.links.model.segnet import segnet_basic
+
+
```
Comment (yuyu2172) [inline examples/segnet/train.py]: Why did you write a train script differently from the style used in Chainer's example scripts?

File: examples/segnet/train.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        t = t[:, ::-1]
+    return x, t
+
+
+train = TransformDataset(train, transform)
+val = CamVidDataset(mode='val')
+
+# Iterator
+train_iter = iterators.MultiprocessIterator(train, args.batchsize)
+val_iter = iterators.MultiprocessIterator(val, args.batchsize, False, False)
```
Comment (yuyu2172) [inline examples/segnet/train.py]: Adding kwargs `shuffle` and `repeat` would be better for readability.

File: examples/segnet/train.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+optimizer.add_hook(chainer.optimizer.WeightDecay(rate=0.0005))
+
+# Updater
+updater = training.StandardUpdater(train_iter, optimizer, device=args.gpu)
+
+
+# Trainer
+trainer = training.Trainer(
+    updater, (140000, 'iteration'),
+    out='results/{}'.format(time.strftime('%Y-%m-%d_%H-%M-%S')))
```
Comment (yuyu2172) [inline examples/segnet/train.py]: Why not `result`, which is used in Chainer's examples.

File: examples/segnet/README.md [hunk 20]
```diff
@@ -0,0 +1,39 @@
+SegNet
+======
+
+# Preparation
+
+Please create `class_weight.npy` using calc_weight.py first. Just run:
+
```
Comment (yuyu2172) [inline examples/segnet/README.md]: Could you summarize the difference between the original implementation and the ChainerCV implementation in this README?
Like you did in the PR comment?

**Comment by yuyu2172** [comment]
I think that "classifier" would be a better word to name "loss" links.
This is because its functionality is different from links in `chainer.links.loss`, but similar to `chainer.links.model.classifiers`.

File: chainercv/links/model/segnet/segnet_basic.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+    SegNetBasic model that is found here_.
+
+    .. [#] Vijay Badrinarayanan, Alex Kendall and Roberto Cipolla "SegNet: A \
+    Deep Convolutional Encoder-Decoder Architecture for Image Segmentation." \
+    PAMI, 2017 
+
+    .. _here: http://github.com/alexgkendall/SegNet-Tutorial
+
+    Args:
+        out_ch (int): The number of channels for the final convolutional layer.
```
Comment (yuyu2172) [inline chainercv/links/model/segnet/segnet_basic.py]: It seems that Chainer's links use `out_channel` across the library.
Why not use `out_channel` instead of `out_ch`?

File: examples/segnet/README.md (hunk 20 above)
Comment (yuyu2172) [inline examples/segnet/README.md]: I think you shoud add references to this README.

File: examples/segnet/README.md [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+3: 0.8427641436
+4: 0.607411863769
+5: 0.628452506525
+6: 0.209502780697
+7: 0.241587759845
+8: 0.615301416933
+9: 0.326346484168
+10: 0.302901995467
+mean: 0.504955548036
+```
```
Comment (yuyu2172) [inline examples/segnet/README.md]: Perhaps this can be made more readable by using class names instead of plain indices.

File: chainercv/links/model/segnet/segnet_basic.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+        return F.upsampling_2d(
+            x, pool.indexes, ksize=(pool.kh, pool.kw),
+            stride=(pool.sy, pool.sx), pad=(pool.ph, pool.pw), outsize=outsize)
+
+    def __call__(self, x):
+        p1 = F.MaxPooling2D(2, 2, use_cudnn=False)
+        p2 = F.MaxPooling2D(2, 2, use_cudnn=False)
+        p3 = F.MaxPooling2D(2, 2, use_cudnn=False)
+        p4 = F.MaxPooling2D(2, 2, use_cudnn=False)
+        h = F.local_response_normalization(x, 5, 1, 0.0001, 0.75)
```
Comment (yuyu2172) [inline chainercv/links/model/segnet/segnet_basic.py]: Sorry if I am wrong, but I think the behavior of this function is different from the Caffe implementation.
The definitions of `alpha` in Chainer and Caffe are different.
I think you need to divide `alpha` by filter size (i.e. 5) in this case.

Similar modification is done in GoogLeNet implementation in Chainer. The model uses LRN as well.

https://github.com/pfnet/chainer/blob/master/chainer/links/model/vision/googlenet.py#L386

File: examples/segnet/calc_weight.py [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+n_cls_pixels = [0 for _ in range(n_class)]
+n_img_pixels = [0 for _ in range(n_class)]
+
+for img, lbl in tqdm(d):
+    for cls_i in np.unique(lbl):
+        n_cls_pixels[cls_i] += np.sum(lbl == cls_i)
+        n_img_pixels[cls_i] += lbl.size
+freq = []
+for n_cls_pixel, n_img_pixel in zip(n_cls_pixels, n_img_pixels):
+    freq.append(n_cls_pixel / float(n_img_pixel))
```
Comment (yuyu2172) [inline examples/segnet/calc_weight.py]: How about importing `from __future__ import division`, and remove `float(...)`.

File: examples/segnet/evaluate.py [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+        n_positive[cls_i] += np.sum(y == cls_i)
+        n_true[cls_i] += np.sum(t == cls_i)
+        n_true_positive[cls_i] += np.sum((y == cls_i) * (t == cls_i))
+
+ious = []
+for cls_i in range(args.n_class):
+    if cls_i in args.ignore_labels:
+        continue
+    deno = float(n_positive[cls_i] + n_true[cls_i] - n_true_positive[cls_i])
+    iou = n_true_positive[cls_i] / deno
```
Comment (yuyu2172) [inline examples/segnet/evaluate.py]: How about importing `from __future__ import division`, and remove `float(...)`.

**Comment by yuyu2172** [comment]
Can you add `camvid_label_names` and `camvid_label_colors`?

File: chainercv/links/model/segnet/segnet_basic.py (hunk 23 above)
Comment (mitmul) [inline chainercv/links/model/segnet/segnet_basic.py]: Thank you very much, I fixed it.

File: examples/segnet/calc_weight.py (hunk 24 above)
Comment (mitmul) [inline examples/segnet/calc_weight.py]: Thank you very much, I fixed it.

File: examples/segnet/evaluate.py (hunk 25 above)
Comment (mitmul) [inline examples/segnet/evaluate.py]: Thank you very much, I fixed it.

File: chainercv/links/model/segnet/segnet_basic.py (hunk 21 above)
Comment (mitmul) [inline chainercv/links/model/segnet/segnet_basic.py]: Thank you very much, I fixed it.

File: examples/segnet/train.py (hunk 19 above)
Comment (mitmul) [inline examples/segnet/train.py]: Thank you very much, I fixed it.

File: examples/segnet/train.py (hunk 18 above)
Comment (mitmul) [inline examples/segnet/train.py]: Thank you very much, I fixed it.

File: examples/segnet/evaluate.py (hunk 16 above)
Comment (mitmul) [inline examples/segnet/evaluate.py]: Thank you very much, I fixed it.

File: chainercv/datasets/camvid/camvid_dataset.py (hunk 15 above)
Comment (mitmul) [inline chainercv/datasets/camvid/camvid_dataset.py]: Thank you very much, I fixed it.

File: chainercv/datasets/camvid/camvid_dataset.py (hunk 14 above)
Comment (mitmul) [inline chainercv/datasets/camvid/camvid_dataset.py]: So do I need to fix the `PixelwiseSoftmaxClassifier` and `PixelwiseSoftmaxWithWeightClassifier` to change the label image shape to (N, H, W) before calling `F.softmax_cross_entropy`, right?

File: chainercv/links/loss/semantic_segmentation_loss.py (hunk 12 above)
Comment (mitmul) [inline chainercv/links/loss/semantic_segmentation_loss.py]: Thank you very much, I fixed it.

File: chainercv/links/loss/semantic_segmentation_loss.py (hunk 13 above)
Comment (mitmul) [inline chainercv/links/loss/semantic_segmentation_loss.py]: Thank you very much, I fixed it.

File: chainercv/links/loss/semantic_segmentation_loss.py (hunk 11 above)
Comment (mitmul) [inline chainercv/links/loss/semantic_segmentation_loss.py]: Thank you very much, I fixed it.

File: examples/segnet/train.py (hunk 9 above)
Comment (mitmul) [inline examples/segnet/train.py]: Thank you very much, I fixed it.

File: examples/segnet/evaluate.py (hunk 8 above)
Comment (mitmul) [inline examples/segnet/evaluate.py]: Thank you very much, I fixed it.

File: chainercv/links/model/__init__.py (hunk 7 above)
Comment (mitmul) [inline chainercv/links/model/__init__.py]: Thank you very much, I fixed it.

File: chainercv/datasets/camvid/camvid_dataset.py (hunk 6 above)
Comment (mitmul) [inline chainercv/datasets/camvid/camvid_dataset.py]: Thank you very much, I fixed it.

File: chainercv/links/model/segnet/segnet_basic.py (hunk 3 above)
Comment (mitmul) [inline chainercv/links/model/segnet/segnet_basic.py]: Thank you very much, I fixed it.

File: chainercv/datasets/camvid/camvid_dataset.py (hunk 4 above)
Comment (mitmul) [inline chainercv/datasets/camvid/camvid_dataset.py]: Thank you very much, I fixed it.

File: examples/segnet/train.py (hunk 17 above)
Comment (mitmul) [inline examples/segnet/train.py]: @yuyu2172 I just tried to make the training script as simple as possible. Because it doesn't need to change many parameters by giving args to the script because its goal is to replicate the paper's result. Is it unnatural?

Comment (yuyu2172) [inline examples/segnet/train.py]: How about adding an option for output directory and writing scripts in `main`.
I think other options like `resume` are not necessary.

File: chainercv/links/model/segnet/segnet_basic.py (hunk 23 above)
Comment (mitmul) [inline chainercv/links/model/segnet/segnet_basic.py]: memo: https://github.com/alexgkendall/caffe-segnet/blob/segnet-cleaned/src/caffe/layers/lrn_layer.cpp#L121

**Comment by yuyu2172** [comment]
I confirmed that `evaluate.py` produces the score reported in README on both Python2 and Python3.

**Comment by yuyu2172** [comment]
I trained on Python2 and the score was as follows. 

```
                    Sky : 0.8676
               Building : 0.6366
                   Pole : 0.1664
           Road_marking : 0.8396
                   Road : 0.6001
               Pavement : 0.6053
                   Tree : 0.1309
             SignSymbol : 0.2214
              Fence,Car : 0.6237
             Pedestrian : 0.2381
              Bicyclist : 0.1942
==================================
               mean IoU : 0.4658
 Class average accuracy : 0.6705
Global average accuracy : 0.8202

```

It scored less than the score posted by mitmul-san, but it is still higher than the reported score.

**Comment by yuyu2172** [comment]
I added a demo. Please use this image.
0001tp_008550

Output

**Comment by yuyu2172** [comment]
This PR is ready for review.

File: examples/segnet/calc_weight.py [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+from chainercv.datasets import CamVidDataset
+
+
+n_class = 12
+dataset = CamVidDataset(split='train')
+
+n_cls_pixels = [0 for _ in range(n_class)]
+n_img_pixels = [0 for _ in range(n_class)]
+
+for img, lbl in dataset:
```
Comment (Hakuyume) [inline examples/segnet/calc_weight.py]: `lbl` -> `label`

File: examples/segnet/evaluate.sh [hunk 27]
```diff
@@ -0,0 +1,2 @@
+python calc_bn_statistics.py --gpu $1 --snapshot $2
+python evaluate.py --gpu $1 --snapshot $2_infer
```
Comment (Hakuyume) [inline examples/segnet/evaluate.sh]: How about calling `calc_bn_statistics` from `evaluate.py`? I think `*.sh` is not good for windows users.

File: examples/segnet/calc_weight.py [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+    for cls_i in np.unique(lbl):
+        n_cls_pixels[cls_i] += np.sum(lbl == cls_i)
+        n_img_pixels[cls_i] += lbl.size
+freq = []
+for n_cls_pixel, n_img_pixel in zip(n_cls_pixels, n_img_pixels):
+    freq.append(n_cls_pixel / n_img_pixel)
+freq = np.array(freq)
```
Comment (Hakuyume) [inline examples/segnet/calc_weight.py]: If we initialize `n_cls_pixels` and `n_img_pixels` by `numpy.zeros`, we can write this part as follows.
```
n_cls_pixels = np.zeros(n_class)
n_img_pixels = np.zeros(n_class)
...
freq = n_cls_pixels / n_img_pixels
```

File: examples/segnet/evaluate.py [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+    serializers.load_npz(args.snapshot, model)
+    if args.gpu >= 0:
+        model.to_gpu(args.gpu)
+
+    test = CamVidDataset(split='test')
+
+    n_positive = [0 for _ in range(n_class)]
+    n_true = [0 for _ in range(n_class)]
+    n_true_positive = [0 for _ in range(n_class)]
+    for i in range(0, len(test), args.batchsize):
```
Comment (Hakuyume) [inline examples/segnet/evaluate.py]: We can use `chainer.iterators,SerialIterator`.

File: examples/segnet/train.py [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+    validation_trigger = (2000, 'iteration')
+    end_trigger = (16000, 'iteration')
+
+    # Dataset
+    train = CamVidDataset(split='train')
+
+    def transform(in_data):
+        img, label = in_data
+        if np.random.rand() > 0.5:
+            img = img[:, :, ::-1]
```
Comment (Hakuyume) [inline examples/segnet/train.py]: We can use `transfroms.random_flip`. (but current code is enough simple...)

File: examples/segnet/evaluate.py [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+    # This follows evaluation code used in SegNet.
+    # https://github.com/alexgkendall/SegNet-Tutorial/blob/master/
+    # # Scripts/compute_test_results.m
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--gpu', type=int, default=-1)
+    parser.add_argument('--snapshot', type=str)
+    parser.add_argument('--batchsize', type=int, default=24)
+    args = parser.parse_args()
+
+    n_class = 11
```
Comment (Hakuyume) [inline examples/segnet/evaluate.py]: In other file, `n_class = 12` is used. If those have different definitions, it is better to used different names.

Comment (yuyu2172) [inline examples/segnet/evaluate.py]: I think this is an issue too.
Maybe, we should assign -1 to pixels with label 11.

File: examples/segnet/train.py (hunk 30 above)
Comment (yuyu2172) [inline examples/segnet/train.py]: I thought about it, but I found this to be simpler.

File: chainercv/links/model/segnet/segnet_basic.py [hunk 32]
```diff
@@ -0,0 +1,130 @@
...
+    .. [#] Vijay Badrinarayanan, Alex Kendall and Roberto Cipolla "SegNet: A \
+    Deep Convolutional Encoder-Decoder Architecture for Image Segmentation." \
+    PAMI, 2017
+
+    .. _here: http://github.com/alexgkendall/SegNet-Tutorial
+
+    Args:
+        n_class (int): The number of channels for the final convolutional
+            layer. SegNetBasic basically takes the number of target classes as
+            this argment.
```
Comment (yuyu2172) [inline chainercv/links/model/segnet/segnet_basic.py]: I will do it in this PR.

**Comment by yuyu2172** [comment]
Thanks. I reflected comments.

File: chainercv/links/model/pixelwise_softmax_classifier.py [hunk 33]
```diff
@@ -0,0 +1,84 @@
...
+        if self.class_weight is not None:
+            self.class_weight = cuda.to_cpu(self.class_weight)
+
+    def to_gpu(self, device=None):
+        super(PixelwiseSoftmaxClassifier, self).to_gpu(device)
+        if self.class_weight is not None:
+            self.class_weight = cuda.to_gpu(self.class_weight, device)
+
+    def __call__(self, x, t):
+        """Computes the loss value for an input and label pair.
```
Comment (Hakuyume) [inline chainercv/links/model/pixelwise_softmax_classifier.py]: `an input and label pair` -> `an image and label pair`?

File: chainercv/links/model/pixelwise_softmax_classifier.py [hunk 34]
```diff
@@ -0,0 +1,84 @@
...
+
+        It also computes accuracy and stores it to the attribute.
+
+        Args:
+            x (~chainer.Variable): A variable with a 4D image array.
+            t (~chainer.Variable): A variable with the ground truth
+                image-wise label.
```
Comment (Hakuyume) [inline chainercv/links/model/pixelwise_softmax_classifier.py]: What does `a 4D image array` mean? `(B, C, H, W)`?

File: chainercv/links/model/segnet/segnet_basic.py [hunk 35]
```diff
@@ -0,0 +1,184 @@
...
+    .. [#] Vijay Badrinarayanan, Alex Kendall and Roberto Cipolla "SegNet: A \
+    Deep Convolutional Encoder-Decoder Architecture for Image Segmentation." \
+    PAMI, 2017
+
+    .. _here: http://github.com/alexgkendall/SegNet-Tutorial
+
+    Args:
+        n_class (int): The number of classes. If :obj:`None`, it can
+            be infered if :obj:`pretrained_model` is given.
+        pretrained_model (str): The destination of the pre-trained
```
Comment (Hakuyume) [inline chainercv/links/model/segnet/segnet_basic.py]: You used `pretrained` above. (I found same issue in `FasterRCNNVGG16`)

File: chainercv/links/model/pixelwise_softmax_classifier.py (hunk 34 above)
Comment (yuyu2172) [inline chainercv/links/model/pixelwise_softmax_classifier.py]: I will change it to a batch of images.

File: chainercv/datasets/camvid/camvid_dataset.py [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+    (128, 64, 128),
+    (60, 40, 222),
+    (128, 128, 0),
+    (192, 128, 128),
+    (64, 64, 128),
+    (64, 0, 128),
+    (64, 64, 0),
+    (0, 128, 192),
+)
+
```
Comment (yuyu2172) [inline chainercv/datasets/camvid/camvid_dataset.py]: Nice idea.

File: chainercv/links/model/segnet/segnet_basic.py (hunk 35 above)
Comment (yuyu2172) [inline chainercv/links/model/segnet/segnet_basic.py]: I will use `pretrained`.
Thanks.

**Comment by Hakuyume** [comment]
I think this PR is OK.

**Comment by rezoo** [comment]
Thank you for taking a lot of time to review this PR! I briefly read that code and LGTM.
